stl/inc/xutility

```diff
@@ -5295,6 +5373,29 @@ _NODISCARD _CONSTEXPR20 bool _Check_match_counts(
             --_Last1;
             --_Last2;
         } while (_Pred(*_Last1, *_Last2));
+
+        if (_First1 == _Last1) {
+            return false;
+        }
+
+        for (;;) {
+            _TrimResult _Res = _Trim_reversed(_First1, _Last1, _First2, _Last2, _Pred);
+            if (_Res != _TrimResult::_KeepTrimming) {
+                if (_Res == _TrimResult::_Break) {
+                    break;
+                }
+                return _Res == _TrimResult::_ReturnTrue;
+            }
+
+            _Res = _Trim_equal(_First1, _Last1, _First2, _Last2, _Pred);
+            if (_Res != _TrimResult::_KeepTrimming) {
+                if (_Res == _TrimResult::_Break) {
+                    break;
+                }
+                return _Res == _TrimResult::_ReturnTrue;
+            }

```

Comment:
Lines 5383-5388 and 5391-5396 are exactly the same, but it's not immediately obvious on how to reduce this repeated code.  I believe, we could rework the loop so that it only appears once: 
```c++
bool _Check_reversed = true;
for (;;) {
   _TrimResult _Res;
   if (_Check_reversed) {
      _Res = _Trim_reversed(_First1, _Last1, _First2, _Last2, _Pred);
   } else {
      _Res = _Trim_equal(_First1, _Last1, _First2, _Last2, _Pred);
   }

   _Check_reversed = !_Check_reversed; // alternate between _Trim_reversed and _Trim_equal 

   if (_Res != _TrimResult::_KeepTrimming) {
      if (_Res == _TrimResult::_Break) {
         break;
      }

      return _Res == _TrimResult::_ReturnTrue;
   }
}
```
I personally prefer this structure because it only has the nested conditional structure appear once, and I believe this should be equivalent to the existing code.  I do recognize that this doesn't bring any behavior change so again, I'm happy for this to be noted for inclusion in a later clean-up PR if we don't want to reset testing.

Reply:
Agreed - and there were 3 total occurrences of this, so I performed a more significant reworking. Now this logic has been extracted into a `_Trim_completely()` helper, with your suggested `_Check_reversed` alternation. I found it simpler to put logic into the `for`-loop. Also, the helper function allows us to avoid the slightly-strange-looking nested conditionals to inspect `_Res`. Now, `_Trim_completely()` is only concerned with whether it sees `_KeepTrimming`, and returns any other values. Then the callers can inspect the result to determine whether they can immediately return an answer, or if more work is necessary.

Finally, because `_Break` no longer corresponds to `break;` that appeared in the earlier code, I have renamed this enumerator to `_HaveWorkAfterTrimming` (and updated a couple of nearby comments). I thought that this read the most naturally in the callers: if we don't have work after trimming, then we can immediately return an answer. (`_DoneTrimming` was my initial idea, but that reads as "if we're not done trimming" which would be confusing.)

Reply:
I really really like the new structure you've put together!  Thanks for improving upon my suggestion :)